README.md
# log-browser-check

[![Latest Version on Packagist][ico-version]][link-packagist]
[![Software License][ico-license]](LICENSE.md)
[![Build Status][ico-travis]][link-travis]
[![Style CI][ico-styleci]][link-styleci]
[![Code Coverage][ico-code-quality]][link-code-quality]
[![Total Downloads][ico-downloads]][link-downloads]

A simple web server log checker to monitor the browser diversity on websites.

## Structure

```
src/
tests/
vendor/
```

## Install

Via Composer

``` bash
$ composer require pxgamer/log-browser-check
```

## Usage

### Supported logs

Server | Supported?
------ | ----------
IIS    | Yes
Apache | No
Nginx  | No

```php
use pxgamer\LogBrowserChecker\Config;
use pxgamer\LogBrowserChecker\IIS;

$config = new Config([
    'session_column' => 7,
    'ip_column' => 6,
    'root_dir' => __DIR__ . '/logs',
    'site_name' => 'test.com',
    'ignored_ips' => ['127.0.0.1']
]);

$browserCheck = new IIS($config);
$browserCheck->findFiles();
$browserCheck->execute();
		
// Retrieve an array of browsers and their usage counts
$browserCheck->getBrowserStats();

// Retrieve an array of unique session ids
$browserCheck->getSessionsIds();

// Retrieve an array of unique user agents
$browserCheck->getUserAgents();
```

## Change log

Please see [CHANGELOG](CHANGELOG.md) for more information on what has changed recently.

## Testing

``` bash
$ composer test
```

## Contributing

Please see [CONTRIBUTING](CONTRIBUTING.md) and [CODE_OF_CONDUCT](CODE_OF_CONDUCT.md) for details.

## Security

If you discover any security related issues, please email owzie123@gmail.com instead of using the issue tracker.

## Credits

- [pxgamer][link-author]
- [All Contributors][link-contributors]

## License

The MIT License (MIT). Please see [License File](LICENSE.md) for more information.

[ico-version]: https://img.shields.io/packagist/v/pxgamer/log-browser-check.svg?style=flat-square
[ico-license]: https://img.shields.io/badge/license-MIT-brightgreen.svg?style=flat-square
[ico-travis]: https://img.shields.io/travis/pxgamer/log-browser-check/master.svg?style=flat-square
[ico-styleci]: https://styleci.io/repos/69864945/shield
[ico-code-quality]: https://img.shields.io/codecov/c/github/pxgamer/log-browser-check.svg?style=flat-square
[ico-downloads]: https://img.shields.io/packagist/dt/pxgamer/log-browser-check.svg?style=flat-square

[link-packagist]: https://packagist.org/packages/pxgamer/log-browser-check
[link-travis]: https://travis-ci.org/pxgamer/log-browser-check
[link-styleci]: https://styleci.io/repos/69864945
[link-code-quality]: https://codecov.io/gh/pxgamer/log-browser-check
[link-downloads]: https://packagist.org/packages/pxgamer/log-browser-check
[link-author]: https://github.com/pxgamer
[link-contributors]: ../../contributors


CHANGELOG.md
# Changelog

All notable changes to `log-browser-check` will be documented in this file.

Updates should follow the [Keep a CHANGELOG](http://keepachangelog.com/) principles.

## v3.0.2 - 2017-08-17

### Added
- Updated layout of package
- Added initial test
- Added code sniffing, Travis CI and Style CI

## v3.0.1 - 2017-08-17

### Added
- Updated functions for unique data
- Edited the README example

## v3.0.0 - 2017-08-17

### Added
- Added new functionality
- Added a customisable Config class
- Various tweaks


tests/ConfigTest.php
<?php

namespace pxgamer\LogBrowserChecker;

use PHPUnit\Framework\TestCase;

/**
 * Class ConfigTest
 */
class ConfigTest extends TestCase
{
    public function testCanGetConfigValue()
    {
        $config = new Config([
            'test' => 'value',
        ]);

        $this->assertEquals('value', $config->getValue('test'));
    }
}


src/IIS.php
<?php

namespace pxgamer\LogBrowserChecker;

/**
 * Class IIS
 */
class IIS extends WebServer
{
    /**
     * IIS constructor.
     * @param Config $oConfig
     */
    public function __construct(Config $oConfig)
    {
        parent::__construct($oConfig);
        $this->iTotalPercent = 0;

        return $this;
    }

    /**
     * @return $this
     * @throws \Exception
     */
    public function execute()
    {
        foreach ($this->aUserAgents as $cur) {
            $inp = str_replace("+", " ", $cur);
            $browser = $this->oBrowscap->getBrowser($inp);

            if (!isset($this->aBrowsers[$browser->parent])) {
                $this->aBrowsers[$browser->parent] = 1;
            } else {
                $this->aBrowsers[$browser->parent]++;
            }
        }

        $totalBrowsers = array_sum($this->aBrowsers);
        foreach ($this->aBrowsers as $b) {
            $this->iTotalPercent = $this->iTotalPercent + (($b / $totalBrowsers) * 100);
        }

        if ($this->iTotalPercent < 98) {
            throw new \ErrorException(self::ERROR_NOT_TOTAL_PERCENTAGE);
        }

        return $this;
    }

    /**
     * @param \SplFileInfo $fileInfo
     */
    protected function statsRunner(\SplFileInfo $fileInfo)
    {
        parent::statsRunner($fileInfo);

        $currentFile = fopen($this->siteDir . DIRECTORY_SEPARATOR . $fileInfo, "r");

        if ($currentFile) {
            // Read through file
            while (!feof($currentFile)) {
                $m_sLine = fgets($currentFile);

                // Check for site name validity
                if (preg_match('/PHPSESSID=(.*?) /', $m_sLine, $aCookieSections) && substr($m_sLine, 0, 1) !== '#') {
                    $m_aSplitIds = explode(" ", $m_sLine);

                    $sSessionId = $this->getSessionFromCookieString($aCookieSections[1]);

                    if (!in_array(
                        $sSessionId,
                        $this->aSessions
                    ) && !in_array(
                        $m_aSplitIds[$this->oConfig->getValue('ip_column')],
                        $this->aIPs
                    )) {
                        $this->aSessions[] = $sSessionId;
                        $m_sUserStat = $m_aSplitIds[$this->oConfig->getValue('session_column')];
                        $this->aUserAgents[] = $m_sUserStat;
                    }
                }
            }

            fclose($currentFile);
        }
    }
}


showPrevious.php
<?php

if (isset($_REQUEST['site']) && $_REQUEST['site'] !== '')
{
	$site = $_REQUEST['site'];
	?>
	<!DOCTYPE html>
        <html lang="en">
        <head>
            <meta name="viewport" content="width=device-width, initial-scale=1">
            <script src="https://code.jquery.com/jquery-3.1.1.min.js"
                    integrity="sha256-hVVnYaiADRTO2PzUGmuLJr8BLUSjGIZsDYGmIJLv2b8=" crossorigin="anonymous"></script>
            <link rel="stylesheet" href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/css/bootstrap.min.css"
                  integrity="sha384-BVYiiSIFeK1dGmJRAkycuHAHRg32OmUcww7on3RYdg4Va+PmSTsz/K68vbdEjh4u"
                  crossorigin="anonymous">
            <script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js"
                    integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa"
                    crossorigin="anonymous"></script>
        </head>
        <body>
			<div class="container">
				<h1><?=strtoupper($site)?> Reports</h1>
				<ul>
					<?php
					$files = scandir("stats/$site/");
					foreach ($files as $file) {
						if (is_file("stats/$site/$file"))
						{
							echo "<li><a href='stats/$site/$file' target='_blank'>$file</a></li>";
						}
					}
					?>
				</ul>
	<?php
}
else
{
	header('Location: /infoCheck/');
}

src/InfoCheck.php
<?php

namespace pxgamer\IISLogBrowserChecker;

class InfoCheck
{
    const ERROR_NOT_TOTAL_PERCENTAGE = 'Possible error, less than 100% total';

    public $dir;
    public $aIPs;
    public $aSplitIds;
    public $browsers;
    public $aTotalStats;
    public $browserVersions;
    public $aFiles;
    public $sSessions;
    public $totalPercent;
    public $browsersN;
    public $site;
    public $inp;

    public function __construct($args, $site)
    {
        $this->site = isset($site) ? $site : '';
        $this->dir = "./logs/" . $site . "/";
        $this->aIPs = array('127.0.0.1');
        $this->aSplitIds = array();
        $this->aTotalStats = array();
        $this->browsers = array();
        $this->browserVersions = array();
        $this->sSessions = array();
        $this->aFiles = 0;
        $this->totalPercent = 0;
        $this->browsersN = "";

        if ($args !== '') {
            $this->inp = str_replace("+", " ", $args[1]);
        }

        $this->findFiles();

        return $this;
    }

    public function execute()
    {
        $count_ff = array();

        foreach ($this->aTotalStats as $cur) {
            $inp = str_replace("+", " ", $cur);
            $browser = get_browser($inp, true);
            $this->browsers[] = $browser['parent'];
            $this->browserVersions[] = $browser['version'];
        }

        $this->browsersN = array_unique($this->browsers);
        foreach ($this->browsersN as $i) {
            $count_ff[$i] = $this->findDuplicates($this->browsers, $i);
        }

        $this->browsers = array_unique($this->browsers);

        $total = array_sum($count_ff);

        foreach ($this->browsers as $b) {
            echo "<tr><td>$b</td><td>$count_ff[$b]</td><td>" . round((($count_ff[$b] / $total) * 100),
                    2) . "</td></tr>";
            $this->totalPercent = $this->totalPercent + (($count_ff[$b] / $total) * 100);
        }

        if ($this->totalPercent < 98) {
            throw new \ErrorException(self::ERROR_NOT_TOTAL_PERCENTAGE);
        }

        return $this;
    }

    function findFiles()
    {
        $files = scandir($this->dir);
        foreach ($files as $file) {
            if (strpos($file, '.log')) {
                $currentFile = fopen($this->dir . $file, "r");
                if ($currentFile) {
                    // Read through file
                    while (!feof($currentFile)) {
                        $m_sLine = fgets($currentFile);
                        // Check for site name validity
                        if (preg_match('*PHPSESSID=* ', $m_sLine) == true && substr($m_sLine, 0, 1) !== '#') {
                            $m_aSplitIds = explode(" ", $m_sLine);
                            $sSID = explode("PHPSESSID=", $m_aSplitIds[13]);
                            $sSID = $sSID[1];
                            $sSID = current(explode(';', $sSID));
                            $sSID = current(explode(',', $sSID));
                            if (!in_array($sSID, $this->sSessions) && !in_array($m_aSplitIds[10], $this->aIPs)) {
                                $this->sSessions[] = $sSID;
                                $m_sUserStat = $m_aSplitIds[12];
                                $this->aTotalStats[] = $m_sUserStat;
                            }
                        }
                    }
                }
                fclose($currentFile);
                $this->aFiles = $this->aFiles++;
            }
        }
        return true;
    }

    function findDuplicates($data, $dupval)
    {
        $nb = 0;
        foreach ($data as $key => $val) {
            if (strpos($val, $dupval) !== false) {
                $nb++;
            }
        }
        return $nb;
    }
}

src/WebServer.php
<?php

namespace pxgamer\LogBrowserChecker;

use BrowscapPHP\Browscap;

/**
 * Class WebServer
 */
class WebServer
{
    const ERROR_NOT_TOTAL_PERCENTAGE = 'Possible error, less than 100% total';

    /**
     * @var array|null
     */
    protected $aIPs = ['127.0.0.1'];
    /**
     * @var null|string
     */
    protected $rootDir;
    /**
     * @var null|string
     */
    protected $siteName;
    /**
     * @var string
     */
    protected $siteDir;
    /**
     * @var array
     */
    protected $aUserAgents;
    /**
     * @var array
     */
    protected $aBrowsers;
    /**
     * @var array
     */
    protected $aSessions;
    /**
     * @var int
     */
    protected $iTotalPercent;
    /**
     * @var Config
     */
    protected $oConfig;
    /**
     * @var Browscap
     */
    protected $oBrowscap;
    /**
     * @var \DirectoryIterator
     */
    protected $oFileIterator;

    /**
     * WebServer constructor.
     * @param Config $oConfig
     */
    public function __construct(Config $oConfig)
    {
        $this->oConfig = $oConfig;

        $this->rootDir = $this->oConfig->getValue('root_dir') ? $this->oConfig->getValue('root_dir') : getcwd();

        $this->siteName = $this->oConfig->getValue('site_name') ? $this->oConfig->getValue('site_name') : '';

        $this->siteDir = $this->rootDir . DIRECTORY_SEPARATOR . $this->siteName;
        $this->aUserAgents = [];
        $this->aIPs = $this->oConfig->getValue('ignored_ips') ? $this->oConfig->getValue('ignored_ips') : $this->aIPs;

        $this->aBrowsers = [];
        $this->aSessions = [];
        $this->iTotalPercent = 0;

        $this->oBrowscap = new Browscap();
    }

    /**
     * Look for files to execute on
     *
     * @return $this
     */
    public function findFiles()
    {
        $this->oFileIterator = new \DirectoryIterator($this->siteDir);

        foreach ($this->oFileIterator as $file) {
            if ($file->getExtension() == 'log') {
                $this->statsRunner($file);
            }
        }

        return $this;
    }

    /**
     * Return an array of browsers and their count in format [browser => int]
     *
     * @return array
     */
    public function getBrowserStats()
    {
        return $this->aBrowsers;
    }

    /**
     * Return an array of unique session IDs
     *
     * @return array
     */
    public function getSessionsIds()
    {
        $this->aSessions = array_unique($this->aSessions);
        return $this->aSessions;
    }

    /**
     * Return an array of unique user agent strings
     *
     * @return array
     */
    public function getUserAgents()
    {
        $this->aUserAgents = array_unique($this->aUserAgents);
        return $this->aUserAgents;
    }

    /**
     * Empty extendable function
     */
    public function execute()
    {
    }

    /**
     * Empty extendable function, should be provided an instance of \SplFileInfo by the findFiles() function
     *
     * @param \SplFileInfo $fileInfo
     */
    protected function statsRunner(\SplFileInfo $fileInfo)
    {
    }

    /**
     * Strip a session id from a cookie string
     *
     * @param $sCookie
     * @return mixed
     */
    protected function getSessionFromCookieString($sCookie)
    {
        $sCookie = current(explode(';', $sCookie));
        return current(explode(',', $sCookie));
    }
}


index.php
<!DOCTYPE html>
        <html lang="en">
        <head>
            <meta name="viewport" content="width=device-width, initial-scale=1">
            <script src="https://code.jquery.com/jquery-3.1.1.min.js"
                    integrity="sha256-hVVnYaiADRTO2PzUGmuLJr8BLUSjGIZsDYGmIJLv2b8=" crossorigin="anonymous"></script>
            <link rel="stylesheet" href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/css/bootstrap.min.css"
                  integrity="sha384-BVYiiSIFeK1dGmJRAkycuHAHRg32OmUcww7on3RYdg4Va+PmSTsz/K68vbdEjh4u"
                  crossorigin="anonymous">
            <script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js"
                    integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa"
                    crossorigin="anonymous"></script>
        </head>
        <body>
			<div class="container">
				<?php
				if (isset($_GET['site'])) {
					$site = $_GET['site'];
				}

				/**
				 * Class InfoCheck
				 *
				 * Usage:    php infoCheck.php "User Agent String"
				 * Example:  php infoCheck.php "Mozilla/5.0+(Windows+NT+6.1;+Trident/7.0;+rv:11.0)+like+Gecko"
				 *
				 */
				class InfoCheck
				{
					/**
					 * InfoCheck constructor.
					 * @param $args
					 */
					function __construct($args, $site)
					{
							$this->dir = "./logs/" . $site . "/";
							$this->m_aIPs = array('127.0.0.1');
							$this->m_aSplitIds = array();
							$this->m_aTotalStats = array();
							$this->browsers = array();
							$this->browserVersions = array();
							$this->m_sSessions = array();
							$this->aFiles = 0;
							$this->totalPercent = 0;
							$this->browsersN = "";
							if ($args !== '') {
								$this->inp = str_replace("+", " ", $args[1]);
							}
							$this->findFiles();
							$this->start($site);
					}

					/**
					 * @return string
					 */
					function start($site)
					{
						$count_ff = array();


							echo '<table class="table"><tr><th>Browser Type</th><th>Number of Uses</th><th>Percentage (%)</th></tr>';
							foreach ($this->m_aTotalStats as $cur) {
								$inp = str_replace("+", " ", $cur);
								$browser = get_browser($inp, true);
								$this->browsers[] = $browser['parent'];
								$this->browserVersions[] = $browser['version'];
							}

							$this->browsersN = array_unique($this->browsers);
							foreach ($this->browsersN as $i) {
								$count_ff[$i] = $this->findDuplicates($this->browsers, $i);
							}

							$this->browsers = array_unique($this->browsers);

							$total = array_sum($count_ff);

							foreach ($this->browsers as $b) {
								echo "<tr><td>$b</td><td>$count_ff[$b]</td><td>" . round((($count_ff[$b] / $total) * 100), 2) . "</td></tr>";
								$this->totalPercent = $this->totalPercent + (($count_ff[$b] / $total) * 100);
							}
							echo "<tr><td><b>Total:</b></td><td>$total</td><td>$this->totalPercent</td></tr>";
							if ($this->totalPercent < 98) { echo "<b>Possible error, less than 100% total</b>"; }

							$this->createTXT($count_ff, $site, $total);

					}

					/**
					 * @return bool
					 */
					function findFiles()
					{
						$files = scandir($this->dir);
						foreach ($files as $file) {
							if (strpos($file, '.log')) {
								$currentFile = fopen($this->dir . $file, "r");
								if ($currentFile) {
									// Read through file
									while (!feof($currentFile)) {
										$m_sLine = fgets($currentFile);
										// Check for sitename validity
										if (preg_match('*PHPSESSID=* ', $m_sLine) == true && substr($m_sLine, 0, 1) !== '#') {
											$m_aSplitIds = explode(" ", $m_sLine);
											$sSID = explode("PHPSESSID=", $m_aSplitIds[13]);
											$sSID = $sSID[1];
											$sSID = current(explode(';', $sSID));
											$sSID = current(explode(',', $sSID));
											if (!in_array($sSID, $this->m_sSessions) && !in_array($m_aSplitIds[10], $this->m_aIPs)) {
												$this->m_sSessions[] = $sSID;
												$m_sUserStat = $m_aSplitIds[12];
												$this->m_aTotalStats[] = $m_sUserStat;
											}
										}
									}
								}
								fclose($currentFile);
								$this->aFiles = $this->aFiles++;
							}
						}
						return true;
					}

					/**
					 * @param $data
					 * @param $dupval
					 * @return int
					 */
					function findDuplicates($data, $dupval)
					{
						$nb = 0;
						foreach ($data as $key => $val)
							if (strpos($val, $dupval) !== false) $nb++;
						return $nb;
					}

					function createTXT($count_ff, $site, $total)
					{
						$m_oCurTime = date('Y-m-d_his', time());
						$dateData = "<a class='btn btn-default' download href='stats/$site/"  . $m_oCurTime . ".txt'><span class='glyphicon glyphicon-download'></span> Download " . $site . "/"  . $m_oCurTime . "<a/>";
						if (!file_exists("stats/$site/")) {
							mkdir("stats/$site", 0700);
						}
						$m_hEndFile = fopen("stats/$site/" . $m_oCurTime . ".html", "w") or die("Unable to open file!");
						fwrite($m_hEndFile,
						'<!DOCTYPE html>
        <html lang="en">
        <head>
            <meta name="viewport" content="width=device-width, initial-scale=1">
            <script src="https://code.jquery.com/jquery-1.12.4.min.js"
                    integrity="sha256-hVVnYaiADRTO2PzUGmuLJr8BLUSjGIZsDYGmIJLv2b8=" crossorigin="anonymous"></script>
            <link rel="stylesheet" href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/css/bootstrap.min.css"
                  integrity="sha384-BVYiiSIFeK1dGmJRAkycuHAHRg32OmUcww7on3RYdg4Va+PmSTsz/K68vbdEjh4u"
                  crossorigin="anonymous">
            <script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js"
                    integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa"
                    crossorigin="anonymous"></script>
        </head>
        <body>
			<div class="container">
				<table class="table"><tr><th>Browser Type</th><th>Number of Uses</th></tr>'
						);
						foreach ($this->browsers as $b) {
							fwrite($m_hEndFile, '<tr><td>'. $b .'</td><td>'. $count_ff[$b] .'</td></tr>');
						}
						fwrite($m_hEndFile, '<tr><td><b>Total:</b></td><td>'.$total.'</td></tr></table></div></body>');
						fclose($m_hEndFile);
						echo "<br/><br/>" . $dateData . "<br/><br/>";
					}
				}

				if (!isset($argv)) {
					$argv = '';
				}

				if (isset($site) && !empty($site)) {
					echo "<title>$site</title>";
					$infoCheck = new InfoCheck($argv, $site);
				}
				else { ?>
						<title>InfoCheck Home</title>
						<style>
						.hover-bottom {
							padding-bottom: 0px;
							border: transparent;
							border-bottom: 0.01em solid transparent;
							background: transparent;
						}
						.hover-bottom:focus {
							outline: none;
							border-bottom: 1px dotted grey;
						}
						</style>
							<div style='margin-top: 300px; text-align: center;'>
							<form>
								<h1>Please enter a site prefix</h1>
								<br/>
								<div class='form-group'><pre style='background: none; border: none;'>infoCheck/<b>?site=<input required autocomplete='off' type='text' class='hover-bottom' id='site' name='site' placeholder='siteName'/></b></pre></div><div class='form-group'><input class='btn btn-default' type='submit' value='New Report'> <span class='btn btn-default' id='loadPrevious'>Show Previous Reports</span></div></form></div>
					<?php
				}
				?>
				<script>
					$('#loadPrevious').click(
						function(){
							var site = $('#site').val();
							window.open('showPrevious.php?site=' + site);
						}
					);
				</script>
			</div>
		</body>